addons/udes_stock/models/stock_picking.py

```diff
@@ -2598,6 +2598,71 @@ def search_for_pickings(self, picking_type_id, picking_priorities, limit=1, doma
 
         return picking
 
+    def _find_unfulfillable_moves(self, batch_size=1000):
+        """Find unfullfilable order lines due to lack of stock."""
+        # This method is patterned on SaleOrder._find_unfulfillable_order_lines
+        # in udes_sale_stock.  Changes applied there should be considered for
+        # this method too.
+        Location = self.env["stock.location"]
+        MoveLine = self.env["stock.move"]
+        Quant = self.env["stock.quant"]
+
+        # Create empty record sets for SO lines
+        unfulfillable_lines = MoveLine.browse()
+
+        _logger.info("Checking pickings to cancel due to stock shortage")
+        # Get unreserved stock for each product in locations
+        locations = Location.get_available_stock_locations()
+        stock = defaultdict(int)
+        skip_states = ("assigned", "done", "cancel")
+
+        for r, batch in self.batched(size=batch_size):
+            _logger.info("Checking pickings %d-%d", r[0], r[-1])
+            # Cache the needed fields and only the needed fields
+            # This code has to process tens of thousands of sale order lines
+            # and hundreds of thousands of stock moves.
+            # Caching too much is expensive here because of the sheer number of
+            # records processed: Odoo's field loading becomes a bottleneck and
+            # memory usage skyrockets.
+            # Caching too little is also expensive since:
+            #  * cache misses cause unused fields to be loaded in due to
+            #    prefetching, which leads to overcaching, described above
+            #  * cache misses for stock move fields result in hundreds of small
+            #    loads per batch, which is inefficient due to overheads
+            # For non-relational fields, read() is used instead of mapped()
+            # because it allows for loading of specific fields, whereas
+            # mapped() will load in as many fields as it can due to prefetching.
+            # with_context(prefetch_fields=False) could be used with mapped()
+            # but is limited to a single column at a time.
+            batch.mapped("move_lines")
```

Comment:
Q: Does this not warm the cache for the non-relational fields anyway?

Reply:
I think so, probably what we need is `batch.with_context(prefetch_fields=False).mapped("move_lines")`

Reply:
Okay so not worried about ordering for now, it's safer to leave that as is.

I'm still not following why we are doing mapped twice though. What do we achieve from mapped `move_lines`, then mapping again with a `read`?

I get the read is to fetch those specific fields, but why do we need `batch.with_context(prefetch_fields=False).mapped("move_lines")` ?

```
batch.with_context(prefetch_fields=False).mapped("move_lines")
batch.mapped("move_lines").read(["product_id", "product_uom_qty", "state"], load="_classic_write")
```